### DISCUSSION TITLE
Combined scans

### DISCUSSION THREAD

**Discussion by mengxuanjia** [mobiusklein/ms_deisotope#20] [Ideas]
Hi Joshua,

I'm trying to use ms_deisotope to analyze my LCMS data. My gradient is too long so I want to bin every 10 (or whatever number that makes sense) MS1 scans to save the data size. Is there a function in your scripts that can combine certain scans? I directly use the Thermo raw files. 

Thank you very much.

Mengxuan

**Comment by mobiusklein** [comment]
Hello,

This is relatively straight-forward to do, depending upon how much metadata you want to carry forwards or how much other processing you want to do along the way. You can merge scans together using `Scan.average_with()` [source](https://github.com/mobiusklein/ms_deisotope/blob/master/ms_deisotope/data_source/scan/scan.py#L829-L885). `Scan.average_with` takes a list of `Scan` objects and an m/z grid spacing and returns a new copy of the current scan, replacing `Scan.arrays` with a new array by averaging all the signal from the original scan and the list of scans together. The m/z grid spacing argument is optional, and and the default guess is "okay", but for Orbitrap data you'll get the best fidelity with a value between 0.001 or 0.0001.

```python
import ms_deisotope

reader = ms_deisotope.MSFileLoader("path/to/file")

bin_size = 10
half_size = bin_size // 2

mz_resolution = 0.001
ms1_buffer = []
msn_buffer = []

for precursor, products in enumerate(reader):
    if len(ms1_buffer) == bin_size :
       ref_scan = ms1_buffer[0]
       merged_scan = ref_scan.average_with(ms1_buffer[1:], mz_resolution)
       # if you want to use the sum instead of the mean
       merged_scan.arrays *= bin_size

       do_something(merged_scan, msn_buffer)

       ms1_buffer = ms1_buffer[half_size:]
       msn_buffer = []
    ms1_buffer.append(precursor)
    msn_buffer.extend(products)
if ms1_buffer:
   ref_scan = ms1_buffer[0]
   merged_scan = ref_scan.average_with(ms1_buffer[1:], mz_resolution)
   # if you want to use the sum instead of the mean
   merged_scan.arrays *= bin_size
   
   do_something(merged_scan, msn_buffer)

   ms1_buffer = ms1_buffer[half_size:]
   msn_buffer = []
```

If your MS1 spectra were acquired pre-centroided, this will reprofile them automatically, so you may want to re-pick peaks afterwards. The averaged MS1 spectra will have a **lot** of data points so they will consume more memory than normal.

All the metadata in the MSn scans will still point to their real parent scans, which can be fixed by doing something like this:
```python
def redirect_precursor_ids(products, precursor):
    level_to_redirect = precursor.ms_level + 1
    # Don't re-direct MS3+ scans
    for product in products:
        if product.ms_level == level_to_redirect:
            product.precursor_information.precursor_scan_id = precursor.id
    return products
```

Does this do what you are looking for?

**Reply by mengxuanjia** [reply]
Yes. I will test it.
Thank you very much!
